### DISCUSSION TITLE
How to manage forks for parallel PR's

### DISCUSSION THREAD

**Discussion by blacknell** [arendst/Tasmota#19481] [Q&A] [answered]
Github limits us to a single fork of the master repo. Can anyone advise how to workaround this, perhaps with multiple clones and then a PR from these into the dev stream. This is beyond my Git knowledge.

Without it, I have to wait until the PR is merged before submitting another PR

**Comment by barbudor** [comment, marked as answer]
You should use a different branch for each PR you are working on

In my fork, the "development" branch is always more or less in sync with tasmota's "development"
I never work on the development branch, it always stay as a erference

When I want to work on a new PR, I create a new branch from "development" and start working on it.
On regular base, I'm commiting my local changes to my local branch, optionally pushing it to the same branch in Github

If I need to work on another PR, I make sure that all thee current changes are commited, then I switch back to "development" that I can re-sync to tasmota, and create a new branch

You can have up to as many branches as you want like this
At some point you may want to merge the latest development branch into any of your working branch to keep on following with tasmota nad reduce the chance of a merge clash preparing your PR

When you are ready to PR, insure all your changes on that branch are commited and the branch pushed to your fork in github

Then submit the PR from that branch to tasmota/development

You can still work on other branches at the same time, and even have multiple PR from different branches

**Reply by blacknell** [reply]
thx very much. basically just like when i work on my own repos

**Reply by barbudor** [reply]
Absolutely, no reason to behave differently

You have two ways to keep your reference "development" branch in your fork on par with Tasmota:

1) From Github, you can merge the changes from Tasmota/development into your development, and then pull that on your local computer

2) Add another source to your local git repo using
```
git remote add upstream github.com:arendst/tasmota
```
Now your local repo will be connected to 2 remote repos:
- origin : your fork on github
- upstream : arendst/tasmota

To update your local "development" branch:
```
git checkout development
git pull upstream development
git push
```
The last line will push back to your origin/development to keep it in sync

**Comment by blacknell** [comment]
I'm getting the following from `git pull upstream development `

Note I tried both `arendst/tasmota` and `arendst/Tasmota` from your instructions
```
paul@github.com: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

```
git remote -v show              
origin	https://github.com/blacknell/Tasmota.git (fetch)
origin	https://github.com/blacknell/Tasmota.git (push)
upstream	github.com:arendst/Tasmota (fetch)
upstream	github.com:arendst/Tasmota (push)
```

So I think above should be `git remote add upstream git@github.com:arendst/Tasmota`?

**Reply by barbudor** [reply]
I may have been too quick on the `add remote` command line because I have some shortcuts in my .ssh/config files
Your command looks like more reasonable
